File: Reqnroll/Chapter20/20.3-ParameterizedExecutionPattern/Before/WIMP.Specs/Support/Hooks.cs
using Reqnroll;

using WIMP.App.Data;
using WIMP.Specs.Drivers;

namespace WIMP.Specs.Support;

[Binding]
public class Hooks(AppHostingContext appHostingContext, IDataRepository dataRepository, PaymentGatewaySimulator paymentGatewaySimulator, IDatabaseDriver databaseDriver)
{
    [BeforeScenario(Order = 0)]
    public async Task ResetDatabase()
    {
        await databaseDriver.RecreateDatabase();
    }

    [BeforeScenario("@payment_gateway", Order = 1)]
    public void InitializePaymentGateway()
    {
        appHostingContext.PaymentGateway = paymentGatewaySimulator.Start();
    }

    [BeforeScenario(Order = 2)]
    public void CreateAppHost()
    {
        appHostingContext.AppHost = new WimpAppHost(dataRepository, appHostingContext.PaymentGateway);
    }

    [AfterScenario]
    public void DisposeAppHost()
    {
        appHostingContext.AppHost?.Dispose();
    }

    [AfterScenario("@payment_gateway")]
    public void DisposePaymentGateway()
    {
        appHostingContext.PaymentGateway?.Stop();
    }
}

File: CucumberJVM/Chapter20/20.2-TestConfigurationProviderPattern/After/src/test/java/com/wimp/app/specs/support/Hooks.java
package com.wimp.app.specs.support;

import com.wimp.app.specs.drivers.PaymentGatewayScenarioContextProxy;
import com.wimp.app.specs.drivers.PaymentGatewaySimulator;
import io.cucumber.java.*;

public class Hooks {
    private final PaymentGatewaySimulator paymentGatewaySimulator;
    private final PaymentGatewayScenarioContextProxy paymentGatewayScenarioContextProxy;

    public Hooks(PaymentGatewaySimulator paymentGatewaySimulator, PaymentGatewayScenarioContextProxy paymentGatewayScenarioContextProxy) {
        this.paymentGatewaySimulator = paymentGatewaySimulator;
        this.paymentGatewayScenarioContextProxy = paymentGatewayScenarioContextProxy;
    }

    @Before(value = "@payment_gateway", order = 1)
    public void initializePaymentGateway() {
        var paymentGateway = paymentGatewaySimulator.start();
        paymentGatewayScenarioContextProxy.setPaymentGatewayOverride(paymentGateway);
    }
}
